presto-main/src/main/java/io/prestosql/operator/PagesHash.java

```diff
@@ -168,6 +181,48 @@ public int getAddressIndex(int rightPosition, Page hashChannelsPage, long rawHas
         return -1;
     }
 
+    /**
+     * return array passed as an argument for performance reasons
+     */
+    public void getAddressIndex(int firstRightPosition, int rightPositions, Page hashChannelsPage, long[] rawHashes, long[] result)
+    {
+        verify(rightPositions == rawHashes.length, "Number of positions must match number of hashes");
+
+        int[] pos = new int[rightPositions];
+        for (int i = 0; i < rightPositions; i++) {
+            pos[i] = getHashPosition(rawHashes[i], mask);
+        }
+
+        /*
+         * Instead of looking for a match on a per-position basis, we execute all first tries "at once" in the inner loop.
+         * That way most of the operations that trigger a cache miss will be grouped together and memory prefetch
+         * latencies will overlap with each other.
+         * In the first iteration we resolve most cases, leaving the rest for subsequent iterations.
+         */
+        int found = 0;
+        while (found != rightPositions) {
```

Comment:
as discussed please try to simplify this code:
1. to use sparse array
2. (optionally) remove multi-pass
and see if it impacts performance

Reply:
> to use sparse array

Did not work performance-wise
> remove multi-pass

Well... I accidentally made another 2 loops making it 4 altogether.
Another 1/3 performance increase though